In [24]:
%cd /content
!git clone https://github.com/alina021001/CS543_Reproducing_and_Analyzing_SimMIM.git
%cd CS543_Reproducing_and_Analyzing_SimMIM

/content
fatal: destination path 'CS543_Reproducing_and_Analyzing_SimMIM' already exists and is not an empty directory.
/content/CS543_Reproducing_and_Analyzing_SimMIM


In [25]:
!ls -la

total 48
drwxr-xr-x 8 root root  4096 Apr 22 02:32  .
drwxr-xr-x 1 root root  4096 Apr 22 02:40  ..
drwxr-xr-x 2 root root  4096 Apr 22 02:32  checkpoints
drwxr-xr-x 2 root root  4096 Apr 22 02:32  docs
drwxr-xr-x 8 root root  4096 Apr 22 02:32  .git
drwxr-xr-x 2 root root  4096 Apr 22 02:32  notebooks
drwxr-xr-x 2 root root  4096 Apr 22 02:32  outputs
-rw-r--r-- 1 root root 10983 Apr 22 02:32 'Project Proposal.docx'
-rw-r--r-- 1 root root    40 Apr 22 02:32  README.md
drwxr-xr-x 2 root root  4096 Apr 22 02:32  scripts


In [26]:
!mkdir -p notebooks

In [27]:
!mkdir -p notebooks
!mkdir -p outputs
!mkdir -p checkpoints
!mkdir -p scripts
!mkdir -p docs

In [28]:
!find . -maxdepth 2

.
./outputs
./scripts
./notebooks
./checkpoints
./docs
./README.md
./Project Proposal.docx
./.git
./.git/branches
./.git/index
./.git/HEAD
./.git/refs
./.git/info
./.git/objects
./.git/packed-refs
./.git/config
./.git/hooks
./.git/description
./.git/logs


In [29]:
%cd /content
!git clone https://github.com/microsoft/SimMIM.git
%cd SimMIM
!find . -maxdepth 2 | head -80

/content
fatal: destination path 'SimMIM' already exists and is not an empty directory.
/content/SimMIM
.
./requirements.txt
./data
./data/data_simmim.py
./data/data_finetune.py
./data/__init__.py
./data/__pycache__
./SECURITY.md
./README.md
./figures
./figures/teaser.jpg
./lr_scheduler.py
./main_simmim.py
./config.py
./.gitignore
./models
./models/swin_transformer.py
./models/build.py
./models/__init__.py
./models/__pycache__
./models/vision_transformer.py
./models/simmim.py
./optimizer.py
./configs
./configs/swin_base__800ep
./configs/swin_large__800ep
./configs/vit_base__800ep
./configs/swin_base__100ep
./CODE_OF_CONDUCT.md
./__pycache__
./__pycache__/optimizer.cpython-312.pyc
./__pycache__/config.cpython-312.pyc
./__pycache__/utils.cpython-312.pyc
./__pycache__/logger.cpython-312.pyc
./__pycache__/lr_scheduler.cpython-312.pyc
./utils.py
./SUPPORT.md
./LICENSE
./logger.py
./main_finetune.py
./.git
./.git/branches
./.git/index
./.git/HEAD
./.git/refs
./.git/info
./.git/objects
./.git

In [30]:
!sed -n '1,220p' README.md

# SimMIM

By [Zhenda Xie](https://zdaxie.github.io)\*, [Zheng Zhang](https://stupidzz.github.io/)\*, [Yue Cao](http://yue-cao.me)\*, [Yutong Lin](https://github.com/impiga), [Jianmin Bao](https://jianminbao.github.io/), [Zhuliang Yao](https://github.com/Howal), [Qi Dai](https://www.microsoft.com/en-us/research/people/qid/) and [Han Hu](https://ancientmooner.github.io/)\*.

This repo is the official implementation of ["SimMIM: A Simple Framework for Masked Image Modeling"](https://arxiv.org/abs/2111.09886).

## Updates

***09/29/2022***

SimMIM was merged to [Swin Transformer repo on GitHub](https://github.com/microsoft/Swin-Transformer).

***03/02/2022***

SimMIM got accepted by CVPR 2022. SimMIM was used in ["Swin Transformer V2"](https://github.com/microsoft/Swin-Transformer) to alleviate the data hungry problem for large-scale vision model training.

***12/09/2021***

Initial commits:

1. Pre-trained and fine-tuned models on ImageNet-1K (`Swin Base`, `Swin Large`, and `ViT Base`) ar

In [31]:
!pip install -r requirements.txt

In [32]:
import torch
print(torch.__version__)
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0))

2.10.0+cu128
True
Tesla T4


In [33]:
!nvidia-smi

Wed Apr 22 02:43:16 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   39C    P8             10W /   70W |       3MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [34]:
# test
!sed -n '1,220p' configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml

MODEL:
  TYPE: swin
  NAME: simmim_pretrain
  DROP_PATH_RATE: 0.0
  SWIN:
    EMBED_DIM: 128
    DEPTHS: [ 2, 2, 18, 2 ]
    NUM_HEADS: [ 4, 8, 16, 32 ]
    WINDOW_SIZE: 6
DATA:
  IMG_SIZE: 192
  MASK_PATCH_SIZE: 32
  MASK_RATIO: 0.6
TRAIN:
  EPOCHS: 100
  WARMUP_EPOCHS: 10
  BASE_LR: 2e-4
  WARMUP_LR: 1e-6
  MIN_LR: 1e-5
  WEIGHT_DECAY: 0.05
PRINT_FREQ: 100
SAVE_FREQ: 5
TAG: simmim_pretrain__swin_base__img192_window6__100ep

In [35]:
!grep -n "swin_base__100ep" -n README.md

41:| Swin-Base | 100 | 192x192 | 192x192 | 82.8 | [google](https://drive.google.com/file/d/1Wcbr66JL26FF30Kip9fZa_0lXrDAKP-d/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml) | [google](https://drive.google.com/file/d/1RsgHfjB4B1ZYblXEQVT-FPX3WSvBrxcs/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_finetune__swin_base__img192_window6__100ep.yaml) |
42:| Swin-Base | 100 | 192x192 | 224x224 | 83.5 | [google](https://drive.google.com/file/d/1Wcbr66JL26FF30Kip9fZa_0lXrDAKP-d/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml) | [google](https://drive.google.com/file/d/1mb43BkW56F5smwiX-g7QUUD7f1Rftq8u/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_finetune__swin_base__img224_window7__100ep.yaml) |


In [36]:
!sed -n '40,120p' README.md

| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| Swin-Base | 100 | 192x192 | 192x192 | 82.8 | [google](https://drive.google.com/file/d/1Wcbr66JL26FF30Kip9fZa_0lXrDAKP-d/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml) | [google](https://drive.google.com/file/d/1RsgHfjB4B1ZYblXEQVT-FPX3WSvBrxcs/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_finetune__swin_base__img192_window6__100ep.yaml) |
| Swin-Base | 100 | 192x192 | 224x224 | 83.5 | [google](https://drive.google.com/file/d/1Wcbr66JL26FF30Kip9fZa_0lXrDAKP-d/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml) | [google](https://drive.google.com/file/d/1mb43BkW56F5smwiX-g7QUUD7f1Rftq8u/view?usp=sharing)/[config](configs/swin_base__100ep/simmim_finetune__swin_base__img224_window7__100ep.yaml) |
| Swin-Base | 800 | 192x192 | 224x224 | 84.0 | [google](https://drive.google.com/file/d/15zENvGjHlM71uKQ3d2F

In [47]:
!sed -n '1,220p' main_simmim.py

# --------------------------------------------------------
# SimMIM
# Copyright (c) 2021 Microsoft
# Licensed under The MIT License [see LICENSE for details]
# Written by Ze Liu
# Modified by Zhenda Xie
# --------------------------------------------------------

import os
import time
import argparse
import datetime
import numpy as np

import torch
import torch.backends.cudnn as cudnn
import torch.distributed as dist
from timm.utils import AverageMeter

from config import get_config
from models import build_model
from data import build_loader
from lr_scheduler import build_scheduler
from optimizer import build_optimizer
from logger import create_logger
from utils import load_checkpoint, save_checkpoint, get_grad_norm, auto_resume_helper

try:
    # noinspection PyUnresolvedReferences
    from apex import amp
except ImportError:
    amp = None


def parse_option():
    parser = argparse.ArgumentParser('SimMIM pre-training script', add_help=False)
    parser.add_argument('--cfg', type=str

In [48]:
#start
%cd /content
!mkdir -p tiny_imagenet_like/train/dummy

/content


In [49]:
!wget -O /content/tiny_imagenet_like/train/dummy/img1.jpg https://picsum.photos/256
!wget -O /content/tiny_imagenet_like/train/dummy/img2.jpg https://picsum.photos/257
!wget -O /content/tiny_imagenet_like/train/dummy/img3.jpg https://picsum.photos/258
!ls -R /content/tiny_imagenet_like/train

--2026-04-22 02:45:59--  https://picsum.photos/256
Resolving picsum.photos (picsum.photos)... 104.26.4.30, 172.67.74.163, 104.26.5.30, ...
Connecting to picsum.photos (picsum.photos)|104.26.4.30|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://fastly.picsum.photos/id/383/256/256.jpg?hmac=okxqX7FpgBht8h_E9sAW34QKReOhGnuOxTH6kpCTZLo [following]
--2026-04-22 02:45:59--  https://fastly.picsum.photos/id/383/256/256.jpg?hmac=okxqX7FpgBht8h_E9sAW34QKReOhGnuOxTH6kpCTZLo
Resolving fastly.picsum.photos (fastly.picsum.photos)... 151.101.1.91, 151.101.65.91, 151.101.129.91, ...
Connecting to fastly.picsum.photos (fastly.picsum.photos)|151.101.1.91|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4119 (4.0K) [image/jpeg]
Saving to: ‘/content/tiny_imagenet_like/train/dummy/img1.jpg’

/content/tiny_image 100%[===================>]   4.02K  --.-KB/s    in 0s      

2026-04-22 02:46:00 (62.8 MB/s) - ‘/content/tiny_imagenet_like/train/dummy/

In [50]:
!python -m torch.distributed.launch --help | head -20

/usr/local/lib/python3.12/dist-packages/torch/distributed/launch.py:207: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

  main()
usage: launch.py [-h] [--nnodes NNODES] [--nproc-per-node NPROC_PER_NODE]
                 [--rdzv-backend RDZV_BACKEND] [--rdzv-endpoint RDZV_ENDPOINT]
                 [--rdzv-id RDZV_ID] [--rdzv-conf RDZV_CONF] [--standalone]
                 [--max-restarts MAX_RESTARTS]
                 [--monitor-interval MONITOR_INTERVAL]
                 [--start-method {spawn,fork,forkserver}]
                 [--event-log-handler EVENT_LOG_HANDLER] [--role ROLE] [-m]
                 [--no-python] [--run-path] [--log-dir LOG_DIR] [-r REDI

In [52]:
%cd /content/SimMIM

!python -m torch.distributed.launch --nproc_per_node 1 main_simmim.py \
  --cfg configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml \
  --data-path /content/tiny_imagenet_like/train \
  --batch-size 2 \
  --amp-opt-level O0 \
  --output /content/simmim_test_output \
  --tag smoke_test \
  --opts TRAIN.EPOCHS 1 TRAIN.WARMUP_EPOCHS 0 PRINT_FREQ 1 SAVE_FREQ 1

/content/SimMIM
/usr/local/lib/python3.12/dist-packages/torch/distributed/launch.py:207: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

  main()
=> merge config from configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml
RANK and WORLD_SIZE in environ: 0/1
/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py:83: UserWarning: barrier(): using the device under current context. You can specify `device_id` in `init_process_group` to mute this warning.
  return func(*args, **kwargs)
[rank0]:[W422 02:46:23.999906275 ProcessGroupNCCL.cpp:5138] Guessing device ID based on global rank. This can cause a hang if rank to GPU 

In [53]:
%cd /content/SimMIM

!python -m torch.distributed.launch --nproc_per_node 1 main_simmim.py \
  --cfg configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml \
  --data-path /content/tiny_imagenet_like/train \
  --batch-size 2 \
  --amp-opt-level O0 \
  --output /content/simmim_maskratio_output \
  --tag mask025 \
  --opts TRAIN.EPOCHS 1 TRAIN.WARMUP_EPOCHS 0 PRINT_FREQ 1 SAVE_FREQ 1 DATA.NUM_WORKERS 2 DATA.MASK_RATIO 0.25

/content/SimMIM
/usr/local/lib/python3.12/dist-packages/torch/distributed/launch.py:207: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

  main()
=> merge config from configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml
RANK and WORLD_SIZE in environ: 0/1
/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py:83: UserWarning: barrier(): using the device under current context. You can specify `device_id` in `init_process_group` to mute this warning.
  return func(*args, **kwargs)
[rank0]:[W422 03:01:08.973158290 ProcessGroupNCCL.cpp:5138] Guessing device ID based on global rank. This can cause a hang if rank to GPU 

In [54]:
!python -m torch.distributed.launch --nproc_per_node 1 main_simmim.py \
  --cfg configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml \
  --data-path /content/tiny_imagenet_like/train \
  --batch-size 2 \
  --amp-opt-level O0 \
  --output /content/simmim_maskratio_output \
  --tag mask050 \
  --opts TRAIN.EPOCHS 1 TRAIN.WARMUP_EPOCHS 0 PRINT_FREQ 1 SAVE_FREQ 1 DATA.NUM_WORKERS 2 DATA.MASK_RATIO 0.50

/usr/local/lib/python3.12/dist-packages/torch/distributed/launch.py:207: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

  main()
=> merge config from configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml
RANK and WORLD_SIZE in environ: 0/1
/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py:83: UserWarning: barrier(): using the device under current context. You can specify `device_id` in `init_process_group` to mute this warning.
  return func(*args, **kwargs)
[rank0]:[W422 03:02:41.251808045 ProcessGroupNCCL.cpp:5138] Guessing device ID based on global rank. This can cause a hang if rank to GPU mapping is heter

In [55]:
!python -m torch.distributed.launch --nproc_per_node 1 main_simmim.py \
  --cfg configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml \
  --data-path /content/tiny_imagenet_like/train \
  --batch-size 2 \
  --amp-opt-level O0 \
  --output /content/simmim_maskratio_output \
  --tag mask075 \
  --opts TRAIN.EPOCHS 1 TRAIN.WARMUP_EPOCHS 0 PRINT_FREQ 1 SAVE_FREQ 1 DATA.NUM_WORKERS 2 DATA.MASK_RATIO 0.75

/usr/local/lib/python3.12/dist-packages/torch/distributed/launch.py:207: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See 
https://pytorch.org/docs/stable/distributed.html#launch-utility for 
further instructions

  main()
=> merge config from configs/swin_base__100ep/simmim_pretrain__swin_base__img192_window6__100ep.yaml
RANK and WORLD_SIZE in environ: 0/1
/usr/local/lib/python3.12/dist-packages/torch/distributed/c10d_logger.py:83: UserWarning: barrier(): using the device under current context. You can specify `device_id` in `init_process_group` to mute this warning.
  return func(*args, **kwargs)
[rank0]:[W422 03:03:51.913293893 ProcessGroupNCCL.cpp:5138] Guessing device ID based on global rank. This can cause a hang if rank to GPU mapping is heter

In [60]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!cp "/content/drive/MyDrive/CS543/group project/simmim_smoke_test.ipynb" \
"/content/CS543_Reproducing_and_Analyzing_SimMIM/notebooks/"

# Baseline Smoke Test Result

Config:
- Swin Base 100ep config
- IMG_SIZE = 192
- MASK_RATIO = 0.6
- Batch size = 2
- Epochs = 1
- GPU = Tesla T4

Outcome:
- Training launched successfully
- Data loader worked
- Model initialized
- 1 epoch completed
- Checkpoint saved

Observed loss:0.9926

# Initial Mask Ratio Comparison

We ran a small single-GPU smoke-test experiment using the official Swin-based SimMIM configuration on a tiny ImageNet-style dataset, while changing only the mask ratio and keeping other settings fixed.

Settings:
- Config: `simmim_pretrain__swin_base__img192_window6__100ep.yaml`
- Image size: 192
- Batch size: 2
- Epochs: 1
- GPU: Tesla T4
- Data: tiny ImageNet-style sample set

Results:

| Mask Ratio | Loss |
|---|---:|
| 0.25 | 0.9771 |
| 0.50 | 1.0081 |
| 0.75 | 1.0458 |

Observation:
As the mask ratio increases, the training loss also increases. This suggests that heavier masking makes the reconstruction task more difficult, which is consistent with the intuition behind masked image modeling.

## Note

These results are only from a tiny smoke-test setting with a very small dataset and one training epoch, so they should be interpreted as an initial trend rather than a final quantitative conclusion.